### PR TITLE
Logs, and notificacions by email

### PR DESCRIPTION
**Pull request by dimoniet**
| Q             | A
| ------------- | ---
| Bug fix?      | No
| New feature?  | Yes
| BC breaks?    |  No
| Deprecations? | No
| Fixed tickets | https://github.com/deployphp/deployer/issues/658 & https://github.com/deployphp/deployer/issues/394 ??

I've developed a simple logging system and an email notify system, using monolog for both.
I'm able to decide if I want one or another, or both, setting it's parameters in my deploy.php file.

The logger function expects 2 parameters, the name we want to set to the project inside the log, and the path for the log. Both are mandatory

The emailer function expects 2 parameters, the name we want to set to the project inside the log, and the email address ( or addresses in an array ) where we want to send notifications. There are 3 extra methods to configure 'from' address, the subject and the minimum logging level, based in RFC 5424, when the email will be  sent. This extra methods are optional.

It's not a big implementation, just to cover my needs.

```
logger('project_name', 'log_path');

emailer('project_name', 'email@example.com')
	->from('from@example.com')
	->subject('Error deploying!')
	->level(400);
```

Hope this can help someone with their deploy control.

### CODE REVIEW

File: src/Log/LogCollection.php [hunk 1]
```diff
@@ -0,0 +1,17 @@
+<?php
+
+namespace Deployer\Log;
+
+use Deployer\Collection\Collection;
+
+class LogCollection extends Collection
```
Comment (elfet) [inline src/Log/LogCollection.php]: Why do you need a log collection?

File: test/src/Log/LogSenderTest.php [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+namespace Deployer\Log;
+
+class LogSenderTest extends \PHPUnit_Framework_TestCase
+{
+    public function testLogSender()
+    {
+        $app = new LogSender('name', 'address');
+
+        $this->assertTrue(method_exists($app, 'writeLog'), 'Class does not have method writeLog');
```
Comment (elfet) [inline test/src/Log/LogSenderTest.php]: Test must cover more then that.

File: test/src/Log/LogWriterTest.php [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+namespace Deployer\Log;
+
+class LogWriterTest extends \PHPUnit_Framework_TestCase
+{
+    public function testLogWriter()
+    {
+        $app = new LogWriter('name', 'path');
+
+        $this->assertTrue(method_exists($app, 'writeLog'), 'Class does not have method writeLog');
```
Comment (elfet) [inline test/src/Log/LogWriterTest.php]: Redundant tests.

File: src/Executor/Informer.php [hunk 4]
```diff
@@ -28,6 +35,11 @@ class Informer
     public function __construct(OutputWatcher $output)
     {
         $this->output = $output;
+        
+        $deployer = Deployer::get();
+        if ($deployer->logs->has("email")) {
+            $this->sender = Deployer::get()->logs->get("email");
```
Comment (elfet) [inline src/Executor/Informer.php]: You are getting `email` which is `sender` from `logs`. Better to separate concepts.

**Comment by elfet** [comment]
Log functionality is very important and will be in Deployer for sure. But why do you need email?

File: src/Log/LogCollection.php (hunk 1 above)
Comment (dimoniet) [inline src/Log/LogCollection.php]: It's not necessary, but I've seen that every function that is configurable from deployer is going to a collection, I just wanted to follow the same workflow you had.

File: test/src/Log/LogSenderTest.php (hunk 2 above)
Comment (dimoniet) [inline test/src/Log/LogSenderTest.php]: New to testing, sorry for not covering much. I will try to write more tests.

File: test/src/Log/LogWriterTest.php (hunk 3 above)
Comment (dimoniet) [inline test/src/Log/LogWriterTest.php]: Same answer than before.

File: src/Executor/Informer.php (hunk 4 above)
Comment (dimoniet) [inline src/Executor/Informer.php]: Mails take the information from logs, this is why I set them together.

**Comment by dimoniet** [comment]
In fact what I needed was a notification by mail when an error occurs, because I'm deploying from cron. When I was working on it I realised that I needed some kind of logs to see what happened, and I changed from a mailing system to a log system, taking profit of Monolog's mailing feature.

If you have any improvement or change, tell me and I will try to do it.

**Comment by oanhnn** [comment]
@dimoniet I think you should make a PR with logging feature only.
Users can send email by a custom `success` task.

**Comment by unclead** [comment]
For Logging better to use PSR-3 interface instead of strong coupling with Monolog. Monolog implements PSR-3 and we can easy use it as logger by default,  but user also should should be able to customize logger via configuration, e.g. if i want to use my own logger which implements PSR-3.

**Comment by dimoniet** [comment]
I'm doing a new logging system, with all your requirements.

**Comment by elfet** [comment]
Also there no need for separate logger function. You can use set function:
```php
set('log', './deploy.log');
```

**Comment by dimoniet** [comment]
I was thinking in something like that:

```
//functions.php
function logger($name, $value)
{
    $deployer = Deployer::get();

    if ($value instanceof \Closure) {
        $deployer->setLogger(call_user_func($value));
    }
    elseif (is_string($value)) {
        $logger = new Monolog\Logger($name);
        $logger->pushHandler(new \Monolog\Handler\StreamHandler($value));
        $deployer->setLogger($logger);
    }
    else {
        throw new InvalidArgumentException('Log should be a logger object or the name and path for the default log.');
    }
}
```



Here the user has two options, the "easy" one where he can write only a name and a path and deployer will use Monolog, and the "pro" where the user can configure his own logger configuration.

For example:

```
//deploy.php
logger('deploy',function(){
	$logger = new Monolog\Logger('deploy');
    $logger->pushHandler(new \Monolog\Handler\ErrorLogHandler());

    return $logger;
});
```

What do you think about that? Perhaps we can set only one parameter that would be the path, or the logger object, and put a default name to the log entry.

**Comment by dimoniet** [comment]
Using this "configurable" option I could use the email notiication system embedded in Monolog, that is what I need and wanted to do at first :-)

**Comment by dimoniet** [comment]
One more question, do you want to log -vvv every deploy, or you want to log output console only?

**Comment by rlacerda83** [comment]
@dimoniet, do you finished this implementation?
I need to log all the tasks...

**Comment by muratbsts** [comment]
@elfet, Is there a way saving logs on set function currently? I've tried set like below. But it doesn't work.

```php
set('log', './deploy.log');
````

**Comment by elfet** [comment]
Log system was implemented in deployer v4, but there is no simple way of configuring it right now.

**Pull request closed without merging**